maestro-ios-xctest-runner/maestro-driver-iosUITests/Routes/Handlers/ViewHierarchyHandler.swift

```diff
@@ -57,105 +54,110 @@ struct ViewHierarchyHandler: HTTPHandler {
+                // the keyboard and alert hierarchies separately.
+                if let element = element as? XCUIApplication {
+                    let keyboard = try? logger.measure(message: "Fetch keyboard hierarchy") {
+                         try keyboardHierarchy(element)
+                    }
+
+                    let alerts = try? logger.measure(message: "Fetch alert hierarchy", {
+                        try fullScreenAlertHierarchy(element)
+                    })
+
+                    return AXElement(children: [
+                        hierarchy,
+                        keyboard,
+                        alerts
+                    ].compactMap { $0 })
                 }
 
-                let alerts = try? logger.measure(message: "Fetch alert hierarchy", {
-                    try fullScreenAlertHierarchy(element)
-                })
-
-                return AXElement(children: [
-                    hierarchy,
-                    keyboard,
-                    alerts
-                ].compactMap { $0 })
+                return hierarchy
+            } else {
+                logger.error("Snapshot failure, cannot return view hierarchy due to \(message)")
+                return nil
             }
-
-            return hierarchy
+        }
+    }
+    
+    private func getErrorCode(message: String) -> String {
+        if message.contains("Error kAXErrorIllegalArgument getting snapshot for element") {
+            return "illegal-argument-snapshot-failure"
+        } else {
+            return "unknown-snapshot-failure"
         }
     }
 
-    func keyboardHierarchy(_ element: XCUIApplication) throws -> AXElement {
+    private func keyboardHierarchy(_ element: XCUIApplication) throws -> AXElement? {
+        if !element.keyboards.firstMatch.exists {
+            return nil
+        }
+        
         let keyboard = try objcTry {
              element.keyboards.firstMatch
         }
 
         return try elementHierarchy(xcuiElement: keyboard)
     }
 
-    func fullScreenAlertHierarchy(_ element: XCUIApplication) throws -> AXElement {
-        let alerts = try objcTry {
-            element.alerts.allElementsBoundByIndex
+    func fullScreenAlertHierarchy(_ element: XCUIApplication) throws -> AXElement? {
+        if !element.alerts.firstMatch.exists {
+            return nil
         }
-
-        let alertHierarchies = try alerts.map { alert in
-            try elementHierarchy(xcuiElement: alert)
+        
+        let alert = try objcTry {
+            element.alerts.firstMatch
         }
 
-        return AXElement(children: alertHierarchies)
+        return try elementHierarchy(xcuiElement: alert)
     }
 
     let useFirstParentWithMultipleChildren = false
-    func findRecoveryElement(_ element: XCUIElement) -> XCUIElement {
+    private func findRecoveryElement(_ element: XCUIElement) -> XCUIElement {
         if !useFirstParentWithMultipleChildren {
```

Comment:
@berikv not sure but when does `useFirstParentWithMultipleChildren` this flag is true? I just see constant false value

Reply:
Ahh got it 👍